### DISCUSSION TITLE
[0.15] Removing disableBalancer feature

### DISCUSSION THREAD

**Discussion by icebob** [moleculerjs/moleculer#1228] [Ideas]
Hey guys, I need your feedback regarding the next version of Moleculer.

Is there somebody who uses the `disableBalancer: true` ServiceBroker options? If yes, please describe here or in DM, what's the use case and why are you using it in your project. 

The reason is that this feature is a little awkward, hard to maintain and manage to be compatible with all other features, and I'm thinking that I will remove it from the ServiceBroker in the next version.

Some history: When this feature was implemented, the aim was to skip the ServiceBroker built-in balancing mechanism and use the transporter-used message-broker (NATS, RabbitMQ) built-in balancing mechanism to cover the situation when the caller node crashed, the action request stays in the message broker and it can retry the request to all consumer nodes. It means it added more reliable action calls. But we developed the moleculer-channels which solves it better and can be used with more message brokers. So if you need reliable action calls, better to use the moleculer-channels. 

In brief, the moleculer-channels library covers the reason why the "disableBalancer" feature was created. So, I would like to remove it from the ServiceBroker in the next version.

Thanks in advance for your feedbacks!

**Comment by ccampanale** [comment]
Interestingly, I have never disabled Moleculer balancing but I have disabled `registry.preferLocal` which sounds very similar to disabling the balancer. (I'd be curious to understand the differences TBH.)

In my use case I designed a process for a single node in a family of services to receive a initial event which it would fan out into several events which I did not want being handled all by the first node but instead, balanced across all nodes in the family. Disabling `preferLocal` made sense to accomplish this (and worked as I needed) and I'm not sure what the difference would have been with `disableBalancer: true`. Perhaps it would have worked essentially the same? But if that's the case, then there's always `registry.preferLocal` to fall back on if/when `disableBalancer` is removed?

**Reply by icebob** [reply]
The `preferLocal` is different and not affected by the `disableBalancer` feature, however `preferLocal: true` can't work in disabled balancer mode because in this case the built-in balancer doesn't work.

The `preferLocal: true` says to ServiceBroker that it should call the local service instance on the local node if exists (to minimalize the network latency), if not, it can call it on a remote node. If you set it to `false`, it means always balancing every action calls, regardless the service is running on a local node or a remote node.

About `disableBalancer` feature, here you can see some pictures about the difference: https://github.com/moleculerjs/moleculer/pull/72
So in this case the Moleculer uses topic names without nodeID, so ServiceBroker can't say who executes the action calls. The reason to implement it was that if the caller ServiceBroker and the consumer one crashed, the message stayed in the message broker (in this case on RabbitMQ) and the message broker retries the action call on another consumer. So it added a little reliability to the action calls and event emits. But now this reliability is covered with moleculer-channels.

I hope I could clear the differences :)

**Reply by ccampanale** [reply]
Thanks for the clarification; makes sense! With that being said I still feel confident that, so long as you can still control `preferLocal`, `disableBalancer` would be unnecessary even in my example case. 

To your point, if I understand correctly, if you want to let the transport infrastructure handle balancing instead of the broker, `moleculer-channels` would be the correct approach.

**Reply by dowster** [reply]
> The preferLocal is different and not affected by the disableBalancer feature, however preferLocal: true can't work in disabled balancer mode because in this case the built-in balancer doesn't work.

I think I actually ran into this and started working on a patch for it. My issue was related to the debounce added in `0.14.19` (or some version around there). I had service A and service B (depending on A) starting on the same node, but balancer disabled and using NATS transit. The broker would finish starting service A and then start service B, as the dependency was satisfied. But service B would immediately send a request to service A which was not listening. 

This was the commit/branch that I had started working on, but I stopped because I wasn't sure if the change was aligned with the intent behind `disableBalancer` and `preferLocal` -> https://github.com/moleculerjs/moleculer/compare/master...dowster:moleculer:fix/respect-prefer-local-with-balancer-disabled

**Comment by shawnmcknight** [comment]
The only time I've used `disableBalancer` is when I wanted to test locally but ensure that traffic was forced over the transporter. This is mostly to test scenarios where I wanted to ensure that de/serialization didn't cause any issues with anything. If there is another mechanism to force use of the transporter then I don't think I'd have any use case for `disableBalancer`.

**Comment by dowster** [comment]
> But we developed the moleculer-channels which solves it better and can be used with more message brokers. So if you need reliable action calls, better to use the moleculer-channels.
> 
> In brief, the moleculer-channels library covers the reason why the "disableBalancer" feature was created. So, I would like to remove it from the ServiceBroker in the next version.

I don't think I'm following this part. We are using both `disableBalancer` _and_ moleculer channels but for different use cases. We have some dynamic event subscriptions in one service which feed into a NATS JetStream queue (using channels) to grab all these events and put them into a durable queue. Similar to somebody subscribing SQS to an SNS topic in AWS. I'm not exactly sure how I would utilize channels to replace an action implementation with request/response needs, do you have examples of how that works? 

I think we do prefer to keep the balancer disabled as the balancer caused issues for us in the past, and it almost seems more streamlined to still leave the balancer off when using the NATS transit. Perhaps the balancing mechanisms could move more into the transit layer? I think that was part of our issue where if a service had balancer enabled it wouldn't create subscribers for "unbalanced" topic names in NATS. Which meant we couldn't send messages from a service with `disableBalancer=true` to one with `disableBalancer=false`.

**Comment by ujwal-setlur** [comment]
I agree with @dowster. I use ```disableBalancer: true``` to rely on ```rabbitmq``` balancing, and I use ```moleculer-channels``` to support reliable events. How can I do ```response/request``` with ```moleculer-channels```? I do prefer to keep the builtin balancer disabled as well.

**Comment by icebob** [comment]
Based on the feedbacks, I keep the `disableBalancer` feature in the next 0.15 version.
